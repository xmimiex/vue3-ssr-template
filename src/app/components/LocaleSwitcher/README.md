#Locale Switcher

Component for changing locales.

It is based on data from the configuration file.
