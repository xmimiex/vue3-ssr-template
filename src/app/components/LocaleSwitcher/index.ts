export { default } from './LocaleSwitcher.vue'
