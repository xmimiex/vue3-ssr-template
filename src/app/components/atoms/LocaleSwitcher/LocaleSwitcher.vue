<template>
  <div class="locale-switcher">
    <label for="locale-switcher">
      {{ $t('localeSwitcher.title') }}
    </label>
    <select id="locale-switcher" @change="switchLocale($event.target.value)">
      <template v-for="localeOption in localeOptions" :key="localeOption.label">
        <option :value="localeOption.value" :selected="localeOption.selected">
          {{ localeOption.label }}
        </option>
      </template>
    </select>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import { useI18n } from 'vue-i18n'
import useConfig from '@app/stores/config'
import useContext from '@app/stores/context'

interface LocaleOption {
  value: string
  label: string
  selected: boolean
}

export default defineComponent({
  name: 'LocaleSwitcher',
  setup() {
    const { t } = useI18n()
    const availableLocales = computed(() => useConfig().internationalization.languagesForCountry)
    const currentCountry = computed(() => useContext().country)
    const currentLanguage = computed(() => useContext().language)

    const localeOptions = computed<LocaleOption[]>(() =>
      Object.keys(availableLocales.value)
        .flatMap(country =>
          availableLocales.value[country].map(language => ({
            value: `/${country}/${language}`,
            label: `${t('localeSwitcher.countries.' + country)} - ${language}`,
            selected: country === currentCountry.value && language === currentLanguage.value,
          })),
        ),
    )

    const switchLocale = (value: string) => {
      window.location.href = value
    }

    return {
      localeOptions,
      switchLocale,
    }
  },
})
</script>
