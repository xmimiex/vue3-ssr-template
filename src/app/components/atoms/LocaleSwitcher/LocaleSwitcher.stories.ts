import { Meta, Story } from '@storybook/vue3'
import LocaleSwitcher from './LocaleSwitcher.vue'
import useConfig from '@/stores/config'
import useContext from '@/stores/context'

export default {
  title: 'Atoms/LocaleSwitcher',
} as Meta

const Template: Story = () => ({
  components: { LocaleSwitcher },
  setup() {
    const context = useContext()
    context.$patch({
      country: 'FR',
      language: 'fr',
    })

    const config = useConfig()
    config.$patch({
      internationalization: {
        defaultCountry: 'FR',
        defaultLanguage: 'fr',
        languagesForCountry: {
          UK: [ 'en' ],
          FR: [ 'fr' ],
          CH: [ 'fr', 'it', 'en' ],
        },
      },
    })
  },
  template: `
    <LocaleSwitcher />`,
})

export const Default = Template.bind({})
