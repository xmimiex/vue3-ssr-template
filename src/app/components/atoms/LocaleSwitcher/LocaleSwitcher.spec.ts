import LocaleSwitcher from '.'
import { mount } from '@vue/test-utils'
import useConfig from '@/stores/config'
import useContext from '@/stores/context'
import { createTestingPinia } from '@pinia/testing'
import createI18n from '@/i18n'

describe('LocaleSwitcher', () => {

  it('should render component correctly', async () => {
    const pinia = createTestingPinia()

    const context = useContext()
    context.$patch({
      country: 'FR',
      language: 'fr',
    })

    const config = useConfig()
    config.$patch({
      internationalization: {
        defaultCountry: 'FR',
        defaultLanguage: 'fr',
        languagesForCountry: {
          UK: ['en'],
          FR: ['fr'],
          CH: ['fr', 'it', 'en'],
        },
      },
    })
    const i18n = await createI18n()

    const wrapper = mount(LocaleSwitcher, {
      global: {
        plugins: [pinia, i18n],
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
