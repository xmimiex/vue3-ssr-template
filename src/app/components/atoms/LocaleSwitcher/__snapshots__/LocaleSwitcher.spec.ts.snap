// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleSwitcher should render component correctly 1`] = `
<div
  class="locale-switcher"
>
  <label
    for="locale-switcher"
  >
    Changer de r√©gion
  </label>
  <select
    id="locale-switcher"
  >
    
    <option
      value="/UK/en"
    >
      Angleterre - en
    </option>
    <option
      value="/FR/fr"
    >
      France - fr
    </option>
    <option
      value="/CH/fr"
    >
      Suisse - fr
    </option>
    <option
      value="/CH/it"
    >
      Suisse - it
    </option>
    <option
      value="/CH/en"
    >
      Suisse - en
    </option>
    
  </select>
</div>
`;
